### DISCUSSION TITLE
Hold simulation, reposition the aircraft and restart

### DISCUSSION THREAD

**Discussion by CarloBlanch1967** [JSBSim-Team/jsbsim#613] [Support requests]
Hello, 
First of all let me congratulate and thank  You (The Authors, the contributors, the mantainers) for this worderful work.
I was totally invoved in You project.
I'm not a professional programmer, and I'm learning C++ thanks to Your project so please forgive me if my questions are probably stupid.
My Goal is to build a complete simulation project with JSBSim.
First of all I learned how to interact with JSBSim via UDP,  and I was able to send position data to an external Visual generator (starting from P3D, then X-plane, now I'm working with RSI protocol) , then I was able to connect FCS commands to connect the hardware in the loop to  fly in real time.
With UDP connection all is working.
Then I integrated all signals to an external avionics software with FSUIPC.
Now the next step is the API integration.
I compiled the static library and tried to integrate the simulation with FDMExec class.
All seems working good.
But my question is:  how can I send the commands without problems?
I'm using the GetPropertyValue  and SetPropertyValue functions in the loop, to read the data and write FCS commands, and all seems working well, but when I try to reposition the aircraft airborne I have an Floating-point overflow (parameters: 0x00000000, 0x00001928) during Debug.
I'm Using Visual Studio 2019, I compiled the JSBSim library in 32bit.
After many runs, MSVS C++ tells me  the problem in the "FGQuaternion.cpp" line 202 " double q0q2 = q0*q2;" 
Surely the problem is my fault, probably is the altitude that come from the external visual (x-plane in this case)  is a float, but the values seems normal. 

this is the code I use to send the data to JSBSim:
```c++
    FDMExec->SetPropertyValue("ic/h-agl-ft", xpl_agl);   (xpl_agl = 977.058716)  
    FDMExec->SetPropertyValue("ic/h-sl-ft", xpl_elevation);  (xpl_elevation = 3.05552028e-09)
    FDMExec->SetPropertyValue("position/lat-gc-deg", xpl_latitude);
    FDMExec->SetPropertyValue("position/long-gc-deg", xpl_longitude);
    FDMExec->SetPropertyValue("ic/psi-true-deg",xpl_heading);
    FDMExec->SetPropertyValue("ic/theta-deg", xpl_pitch);
```
The problem could be the elevation?
I tried also to reduce the floating point precision without success.

Thanks in advance.

Carlo.

**Comment by seanmcleod** [comment]
Is it just a typo in terms of the latitude and longitude not being an initial condition, i.e. `ic/lat-gc-deg`? And after you've set the initial conditions are executing `FDMExec->RunIC()`?

**Reply by bcoconni** [reply]
As @seanmcleod mentioned, your initial coordinates should be initialized via the properties `ic/lat-gc-deg` for latitude and `ic/long-gc-deg` for longitude.

In addition `ic/h-sl-ft` is not the terrain elevation as I infer from the variable name `xpl_elevation` but the altitude above sea level ("sl" stands for **S**ea **L**evel). As a consequence the second line of code `FDMExec->SetPropertyValue("ic/h-sl-ft", xpl_elevation);` is overwriting the altitude that has been set by the first line of code and this basically sets the aircraft to an initial altitude of 0 ft. [As I explained in another discussion](https://github.com/JSBSim-Team/jsbsim/discussions/610#discussioncomment-2578709), this can result in weird results since the landing gears are unrealistically compressed.

Below are the suggested changes to your code. Let us know if that fixes your problem.
```diff
    FDMExec->SetPropertyValue("ic/h-agl-ft", xpl_agl);   (xpl_agl = 977.058716)  
-    FDMExec->SetPropertyValue("ic/h-sl-ft", xpl_elevation);  (xpl_elevation = 3.05552028e-09)
+    FDMExec->SetPropertyValue("position/terrain-elevation-asl-ft", xpl_elevation);  (xpl_elevation = 3.05552028e-09)
-    FDMExec->SetPropertyValue("position/lat-gc-deg", xpl_latitude);
+    FDMExec->SetPropertyValue("ic/lat-gc-deg", xpl_latitude);
-    FDMExec->SetPropertyValue("position/long-gc-deg", xpl_longitude);
+    FDMExec->SetPropertyValue("ic/long-gc-deg", xpl_longitude);
    FDMExec->SetPropertyValue("ic/psi-true-deg",xpl_heading);
    FDMExec->SetPropertyValue("ic/theta-deg", xpl_pitch);
```

**Comment by CarloBlanch1967** [comment]
Ok, Thanks!

I will try soon.


So, for reposition the aircraft I must  hold the simulation:

FDMExec->Hold();

Update the IC properties with FDMExec->SetPropertyValue  (all of them?)

Restart the simulation with FDMExec->RunIC();


I changed the topic title.

Thanks.

**Comment by CarloBlanch1967** [comment]
Ok, 


First of all, the main problem was an I/O function that I use to send and receive datas from FSUIPC.

So for now, I stopped its execution into the main loop to isolate the issue.

Now I want to concentrate on repositioning.

I've changed some code because Xplane calls "elevation"  the altitude over sea level.
So for repositioning my aircraft I must subtract the agl altitude from "elevation" (now I'm calling it "xpl_altitude" ) to obtain the elevation to pass as property value.
But after running again with RunIC()  now I am experiencing a weird attitude , strange rotation and the aircraft falling down.
Do I must pass all the IC property values?
The values are correct because I also rewrite the init file and works good if I restart the software.


```c++
   FDMExec->Hold();    
    xpl_elevation = xpl_altitude - xpl_agl;
    FDMExec->SetPropertyValue("ic/h-agl-ft", xpl_agl);
    FDMExec->SetPropertyValue("ic/terrain-elevation-ft", xpl_elevation);
    FDMExec->SetPropertyValue("ic/h-sl-ft", xpl_altitude);
    FDMExec->SetPropertyValue("ic/lat-gc-deg", xpl_latitude);
    FDMExec->SetPropertyValue("ic/long-gc-deg", xpl_longitude);
    FDMExec->SetPropertyValue("ic/psi-true-deg",xpl_heading);
    FDMExec->SetPropertyValue("ic/theta-deg", xpl_pitch);
       
    
    FDMExec->RunIC();
```
Thanks Again

**Comment by seanmcleod** [comment]
I'm a bit confused in terms of your overall setup, since there seem to be multiple simulation engines involved, so it's not entirely clear to me which ones are responsible for the flight dynamics, maybe some are only responsible for rendering the outside view graphics?

You mention FSUIPC which is an API for dealing with Microsoft Flight Simulator or Prepar3D, then you mention XPlane another full flight simulator and then JSBSim.

I have experience from my own VSS (Variable Stability System) simulator where I use JSBSim for the flight dynamics etc. and then on each time step I used the SimConnect API etc. to drive Prepar3D's external views. In my setup I also query Prepar3D for terrain elevation which I pass into JSBSim.

**Comment by CarloBlanch1967** [comment]
Sorry, I forgot to explain my project.
I'm using xplane as visual generator,  using udp to send data (with the VEHX protocol).
The fsuipc integration is used to drive an external avionics suite.
In my first release, my software launch jsbsim.exe,  and the other software I need , reading data from jsbsim via udp, sending the position to xplane via udp ,  reading joystick commands via fsuipc.
Everything works fine, but I don't like too mutch this approach , too many udp stream to control.

To increase the transport delay, I 'm trying to integrate the fdm in the main program.

I need xplane as Visual generator,  but  I also need to query elevation data from it to have the same elevation.

**Comment by seanmcleod** [comment]
> I need xplane as Visual generator, but I also need to query elevation data from it to have the same elevation.

Okay, so I do the exact same thing in my VSS sim, I'm just using Prepar3D instead of X-Plane as the visual generator and as the source of the terrain elevation.

So on each simulation time step I use the lat, lon position calculated by JSBSim to set the aircraft's position in Prepar3D, which updates the visuals, I then also receive the ground elevation for that coordinate from Prepar3D on each time step and feed that into JSBSim.

```c++
  // Receive ground elevation from Prepar3D
  case SIMCONNECT_RECV_ID_SIMOBJECT_DATA:
  {
    SIMCONNECT_RECV_SIMOBJECT_DATA* pObjData = (SIMCONNECT_RECV_SIMOBJECT_DATA*)pData;

    if (pObjData->dwRequestID == AIRCRAFT_INFO_REQUEST)
    {
      AIRCRAFT_INFO_DATA* pInfoData = (AIRCRAFT_INFO_DATA*)&pObjData->dwData;
      AircraftInfoData.GroundAltitude = pInfoData->GroundAltitude * 3.28083989;


void copy_to_JSBSim()
{
  // Set ground altitude
  std::shared_ptr<JSBSim::FGPropagate> Propagate = FDMExec->GetPropagate();
  Propagate->SetTerrainElevation(AircraftInfoData.GroundAltitude);

```

**Reply by CarloBlanch1967** [reply]
Good , I will follow this way.
Thanks.

So  , what's the best way to send  and read data to/from fdm?
You use the FgPropagate class.

**Comment by CarloBlanch1967** [comment]
Now my situation:
If I try to hold the simulation , changing position and resume, I get a unstabe position and the aircraft crash into the ground.
If I try to hold the simulation , changing position and RunIC(),  I get a unstabe position and the aircraft crash into the ground.
If I try to hold the simulation, changing position and ResetToInitialConditions , the application crashes in the same way:

Unhandled exception at 0x0009CC91 in JSBTest.exe: 0xC0000091: Floating-point overflow (parameters: 0x00000000, 0x00001928).

``` c++
FGAtmosphere.cpp at row 127 DensityAltitude = CalculateDensityAltitude(Density, altitude);
```
or
``` c++
FGQuaternion.cpp at row 202 double q0q2 = q0*q2;
```

If I restart my exe, the situation is perfect , because JSBsim restart from the init file just generated with the correct values.

How can I restart the simulation from scratch ?

**Comment by seanmcleod** [comment]
> So , what's the best way to send and read data to/from fdm?

I'm not sure if this is the best way, but this is the way I do it in my sim.

```c++
void copy_to_JSBSim()
{
  // Set ground altitude
  std::shared_ptr<JSBSim::FGPropagate> Propagate = FDMExec->GetPropagate();
  Propagate->SetTerrainElevation(AircraftInfoData.GroundAltitude);
....
  std::shared_ptr<JSBSim::FGFCS> FCS = FDMExec->GetFCS();
  
  // DirectInput inputs
  g_Joystick.Poll();

  FCS->SetDeCmd(g_Joystick.Pitch);
  FCS->SetDaCmd(g_Joystick.Roll);
  FCS->SetDrCmd(DeadBand(g_Joystick.Rudder, 0.02));
  FCS->SetThrottleCmd(-1, g_Joystick.Throttle);
  FCS->SetLBrake(g_Joystick.LeftBrake);
  FCS->SetRBrake(g_Joystick.RightBrake);
....
}
```

And..

```c++
void copy_from_JSBSim()
{
....
    // Pump SimConnect message pump
    SimConnect_CallDispatch(hSimConnect, MyDispatchProcSD, NULL);

    // Collect JSBSim data
    std::shared_ptr<JSBSim::FGPropagate> Propagate = FDMExec->GetPropagate();
    std::shared_ptr<JSBSim::FGAuxiliary> Auxiliary = FDMExec->GetAuxiliary();
    std::shared_ptr<JSBSim::FGAtmosphere> Atmosphere = FDMExec->GetAtmosphere();
    std::shared_ptr<JSBSim::FGAccelerations> Accelerations = FDMExec->GetAccelerations();

    AircraftData.Altitude = Propagate->GetAltitudeASL();
    AircraftData.Latitude = Propagate->GetLocation().GetLatitudeDeg();
    AircraftData.Longitude = Propagate->GetLocation().GetLongitudeDeg();

    AircraftData.Bank = radtodeg * Propagate->GetEuler(ePhi) * -1.0;
    AircraftData.Pitch = radtodeg * Propagate->GetEuler(eTht) * -1.0;
    AircraftData.Heading = radtodeg * Propagate->GetEuler(ePsi);

    AircraftData.IAS = Auxiliary->GetVcalibratedKTS();
    AircraftData.VSI = Propagate->GetVel(eDown) * -1.0;

    AircraftData.BodyZVelocity = Propagate->GetUVW(JSBSim::FGJSBBase::eU);
    AircraftData.BodyXVelocity = Propagate->GetUVW(JSBSim::FGJSBBase::eV);
    AircraftData.BodyYVelocity = Propagate->GetUVW(JSBSim::FGJSBBase::eW) * -1.0;

    AircraftData.WorldZVelocity = Propagate->GetVel(JSBSim::FGJSBBase::eNorth);
    AircraftData.WorldXVelocity = Propagate->GetVel(JSBSim::FGJSBBase::eEast);
    AircraftData.WorldYVelocity = Propagate->GetVel(JSBSim::FGJSBBase::eDown) * -1.0;

    AircraftData.BodyZAcceleration = Accelerations->GetUVWdot(JSBSim::FGJSBBase::eU);
    AircraftData.BodyXAcceleration = Accelerations->GetUVWdot(JSBSim::FGJSBBase::eV);
    AircraftData.BodyYAcceleration = Accelerations->GetUVWdot(JSBSim::FGJSBBase::eW) * -1.0;
....
}
```

**Comment by seanmcleod** [comment]
> How can I restart the simulation from scratch ?

I think `ResetToInitialConditions()` should do the job. So set/adjust the initial conditions as required and then call it.

https://github.com/JSBSim-Team/jsbsim/blob/b4ca39909aec304730ea428b0f8de8b61b55457e/src/FGFDMExec.cpp#L687-L704

It also calls `RunIC()` unless specifically told not to.

https://github.com/JSBSim-Team/jsbsim/blob/b4ca39909aec304730ea428b0f8de8b61b55457e/src/FGFDMExec.cpp#L641-L674

Also take a look at the following discussion with regards to resetting JSBSim for multiple runs - https://github.com/JSBSim-Team/jsbsim/discussions/455

**Comment by CarloBlanch1967** [comment]
Ok Sean, Thank You.

Forgive my incompetence about C++ shared pointers ,  but how can I create the Class constructor?

`std::shared_ptr<JSBSim::FGPropagate> Propagate = FDMExec->GetPropagate();`  
give me the error :
'initializing': cannot convert from 'JSBSim::FGAuxiliary *' to 'std::shared_ptr<JSBSim::FGAuxiliary>'	

Thanks.

**Comment by CarloBlanch1967** [comment]
Ok ,
following the instructions of @bcoconni in the discussion #491 , I downloaded the sources from `master` , I rebuilt the library and set the differences between the 2 original `JSBSim.cpp`.
The "Releases" source code v1.1.11 is different from the "master" one.

**Comment by seanmcleod** [comment]
Yes the master branch always consists of the latest code, releases will be older.
